src/scaffoldmaker/meshtypes/meshtype_1d_path1.py

```diff
@@ -103,28 +104,50 @@ def generateBaseMesh(cls, region, options):
             element.setNodesByIdentifier(eft, [ e + 1, e + 2 ])
             elementIdentifier = elementIdentifier + 1
 
-        fm.endChange()
         return []
 
+    @classmethod
+    def smoothPath(cls, region, options, mode : DerivativeScalingMode):
+        x, d1 = extractPathParametersFromRegion(region)[0:2]
+        d1 = smoothCubicHermiteDerivativesLine(x, d1, magnitudeScalingMode=mode)
+        setPathParameters(region, [ Node.VALUE_LABEL_VALUE, Node.VALUE_LABEL_D_DS1 ], [ x, d1 ])
+
+    @classmethod
+    def getInteractiveFunctions(cls):
+        """
+        Supply client with functions for smoothing path parameters.
+        """
+        return [
+            ("Smooth d1 arithmetic", lambda region, options: cls.smoothPath(region, options, DerivativeScalingMode.ARITHMETIC_MEAN)),
+            ("Smooth d1 harmonic", lambda region, options: cls.smoothPath(region, options, DerivativeScalingMode.HARMONIC_MEAN)) ]
+
 
-def extractPathParametersFromRegion(region):
+def extractPathParametersFromRegion(region, groupName=None):
     '''
     Returns parameters of all nodes in region in identifier order.
     Assumes nodes in region have field coordinates (1 to 3 components).
     Currently limited to nodes with exactly value, d_ds1, d_ds2, d2_ds12,
     same as path 1 scaffold.
+    :param groupName: Optional name of Zinc group to get parameters from.
     :return: cx, cd1, cd2, cd12 (all padded with zeroes to 3 components)
     '''
-    fm = region.getFieldmodule()
-    coordinates = fm.findFieldByName('coordinates').castFiniteElement()
+    fieldmodule = region.getFieldmodule()
+    coordinates = fieldmodule.findFieldByName('coordinates').castFiniteElement()
     componentsCount = coordinates.getNumberOfComponents()
     assert componentsCount in [ 1, 2, 3 ], 'extractPathParametersFromRegion.  Invalid coordinates number of components'
-    cache = fm.createFieldcache()
+    cache = fieldmodule.createFieldcache()
     cx = []
     cd1 = []
     cd2 = []
     cd12 = []
-    nodes = fm.findNodesetByFieldDomainType(Field.DOMAIN_TYPE_NODES)
+    nodes = fieldmodule.findNodesetByFieldDomainType(Field.DOMAIN_TYPE_NODES)
+    if groupName:
+        group = fieldmodule.findFieldByName(groupName).castGroup()
+        nodeGroup = group.getFieldNodeGroup(nodes)
+        if nodeGroup.isValid():
+            nodes = nodeGroup.getNodesetGroup()
+        else:
+            print('extractPathParametersFromRegion: missing group "' + groupName + '"')
```

Comment:
Shouldn't be under ```if groupName:``` ?

Reply:
Note sure what you mean. I just wanted a warning if the group was not found (meaning anything from not existing to having no nodes in it). If not group name is specified, it uses all nodes; if not found it also uses all nodes.

Reply:
I think it should be a part of the first if statement.

Reply:
I might be wrong

Reply:
This is what I think it should be

```
    if groupName:
        group = fieldmodule.findFieldByName(groupName).castGroup()
        nodeGroup = group.getFieldNodeGroup(nodes)
        if nodeGroup.isValid():
            nodes = nodeGroup.getNodesetGroup()
    else:
         print('extractPathParametersFromRegion: missing group "' + groupName + '"')
```

Reply:
No, groupName is optional; if you don't supply it (or it is empty/doesn't exist) you get all nodes.
What may not be so obvious is that the NodesetGroup is a derived type, so the 3rd to last line is replacing all nodes with the subset of nodes, but the following code will work either way (because NodesetGroup implements all functions of the parent domain on its subset).

Reply:
Thanks for clearing that up.